core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java

```diff
@@ -121,21 +121,26 @@ public void onNext(ByteBuffer byteBuffer) {
             if (byteBuffer == null) {
                 throw new NullPointerException("Element must not be null");
             }
-            
+
+            if (byteBuffer.remaining() == 0) {
+                // Immediately skip empty buffers
+                subscription.request(1);
+                return;
+            }
+
             writeInProgress = true;
+
             fileChannel.write(byteBuffer, position.get(), byteBuffer, new CompletionHandler<Integer, ByteBuffer>() {
                 @Override
                 public void completed(Integer result, ByteBuffer attachment) {
-                    if (result > 0) {
-                        position.addAndGet(result);
-                        synchronized (FileSubscriber.this) {
-                            if (closeOnLastWrite) {
-                                close();
-                            } else {
-                                subscription.request(1);
-                            }
-                            writeInProgress = false;
+                    position.addAndGet(result);
+                    synchronized (FileSubscriber.this) {
+                        if (closeOnLastWrite) {
+                            close();
+                        } else {
+                            subscription.request(1);
```

Comment:
We don't seem to check that the buffer was written entirely before requesting another one. Seems like an edge case (fine with addressing in a different PR).

Reply:
Do you know what the expectation is when the buffer was not written entirely?

Reply:
Fixed.